File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws IOException {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	
    	while(true) {
    		String str = br.readLine();
    		
    		if(str.equals(".")) break;
    		
    		System.out.println(result(str));
    	}
	
	
	}
	
	public static String result(String s) {
		Stack<Character> stack = new Stack<Character>();
		
		for(int i = 0; i < s.length(); i++) {
			char ch = s.charAt(i);
			
			if(ch == '(' || ch == '[') {
				stack.push(ch);
				
			} else if(ch == ')') {
				if(stack.isEmpty() || stack.peek() != '(') {
					return "no";
				} else {
					stack.pop();
				}
				
			} else if(ch == ']') {
				
				if(stack.isEmpty() || stack.peek() != '[') {
					return "no";
				} else {
					stack.pop();
				}
				
			} 
		}
		
		if(stack.isEmpty()) {
			return "yes";
		} else {
			return "no";
		}
	}
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

string result(string str)
{
	stack<char> s;

	for (int i = 0; i < str.length(); i++) {
		if (str[i] == '(' || str[i] == '[') {
			s.push(str[i]);
		}
		else if (str[i] == ')') {
			if (s.empty() || s.top() != '(') {
				return "no";
			}
			else {
				s.pop();
			}
		}
		else if (str[i] == ']') {
			if (s.empty() || s.top() != '[') {
				return "no";
			}
			else {
				s.pop();
			}
		}
	}

	if (s.empty()) {
		return "yes";
	}
	else {
		return "no";
	}
}

int main()
{	
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	string str;

	while (1) {
		getline(cin, str);

		if (str.length() == 1 || str[0] == '0')
			break;

		cout << result(str) << "\n";
	}

	return 0;
}